pkg/metadata/provider_azure.go

```diff
@@ -0,0 +1,125 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"path"
+	"time"
+)
+
+const (
+	azMetaDataURL     = "http://169.254.169.254/metadata/instance/"
+	azMetadataVersion = "2017-04-02"
+	azUserDataLoc     = "/var/lib/waagent/ovf-env.xml"
+)
+
+// ProviderAzure is the type implementing the Provider interface for Azure
+type ProviderAzure struct {
+}
+
+// NewAzure returns a new ProviderAzure
+func NewAzure() *ProviderAzure {
+	return &ProviderAzure{}
+}
+
+func (p *ProviderAzure) String() string {
+	return "Azure"
+}
+
+// Probe checks if we are running on Azure
+func (p *ProviderAzure) Probe() bool {
+	// Getting the hostname should always work...
+	_, err := azGet(azMetaDataURL + "compute/name")
+	return (err == nil)
+}
+
+// Extract gets both the Azure specific and generic userdata
+func (p *ProviderAzure) Extract() ([]byte, error) {
+	// Get host name. This must not fail
+	hostname, err := azGet(azMetaDataURL + "compute/name")
+	if err != nil {
+		return nil, err
+	}
+	err = ioutil.WriteFile(path.Join(ConfigPath, Hostname), hostname, 0644)
+	if err != nil {
+		return nil, fmt.Errorf("Azure: Failed to write hostname: %s", err)
+	}
+
+	// public ipv4
+	azMetaGet("network/interface/0/ipv4/ipAddress/0/publicIpAddress", "public_ipv4", 0644)
+
+	// private ipv4
+	azMetaGet("network/interface/0/ipv4/ipAddress/0/privateIpAddress", "local_ipv4", 0644)
+
+	// availability zone
+	azMetaGet("compute/location", "availability_zone", 0644)
+
+	// instance type
+	azMetaGet("compute/vmSize", "instance_type", 0644)
+
+	// instance-id
+	azMetaGet("compute/vmId", "instance_id", 0644)
+
+	// local-hostname
+	azMetaGet("compute/name", "local_hostname", 0644)
+
+	// ssh is not available in Azure
+
+	// Generic userdata, available as XML data
+	userData, err := getUserData()
+	if err != nil {
+		log.Printf("Azure: Failed to get user-data: %s", err)
+		// This is not an error
+		return nil, nil
+	}
+	return userData, nil
+}
+
+// getUserData returns the userdata stored on disk
+func getUserData() ([]byte, error) {
+	return ioutil.ReadFile(azUserDataLoc)
```

Comment:
According to the docs the ovf file is actually on a DVD https://docs.microsoft.com/en-us/azure/virtual-machines/linux/agent-user-guide?toc=%2fazure%2fvirtual-machines%2flinux%2ftoc.json

(I really dislike the waagent, as it is huge and tries to do everything at once which is very complicated; we do currently run it on Docker for Azure, but I would like to replace its functions with standalone containers really, and so I don't think that eg this metadata package should require it to be running, indeed it will be quite hard as it needs to run as a service so there is no way you can get this file mounted as `metadata` needs to run as an `onboot` container.)